projects/DifferentialDrive.md

```diff
@@ -0,0 +1,655 @@
+      rightPIDController.calculate(rightEncoder.getVelocity(), 
+      rightSpeed * DriveConstants.MAX_SPEED);
+```
+
+In this part, the PID controllers calculate how much they need to adjust the motor speed to correct any errors. They do this by comparing the actual speed (from the encoders) with the target speed.
+
+Finally, we combine the outputs and send them to the motors:
+
+```java
+      double leftVoltage = leftOutput + leftFeedforward;
+      double rightVoltage = rightOutput + rightFeedforward;
+
+      leftLeader.setVoltage(leftVoltage);
+      rightLeader.setVoltage(rightVoltage);
+```
+
+Here, the final motor voltage is calculated by adding the PID and feedforward outputs together. These voltages are then sent to the motors, controlling the movement of the robot.
+
+### Unit Testing & System Checks
+
+#### Unit Tests
+
+Think of unit tests as a way to check that each small piece of your code—like a motor or sensor—is functioning properly on its own before you start putting them all together. It's much easier to fix problems when they’re isolated to a single component than when they’re all tangled up in the full system.
+
+**Make sure to check out this [guide](https://docs.wpilib.org/en/stable/docs/software/wpilib-tools/robot-simulation/unit-testing.html) from WPILib on how to setup unit tests.**
+
+Let’s talk about the tests for our drivetrain. We'll start it by creating a `DriveTest.java` file in the robot folder of the tests folder.
+
+First, we need to set things up before we run any tests. This is done in the `setup()` method:
+
+```java
+  Drive drive;
+  CANSparkMax spark;
+
+  @BeforeEach
+  public void setup() {
+    setupTests();
+    drive = new Drive();
+  }
+```
+
+Next, we will need to reset our drive between every tests so we are able to experiment different things at once.
+
+```java
+  @AfterEach
+  void destroy() throws Exception {
+    reset(drive);
+  }
+```
+
+Once everything is set up, you can make tests with the `@Test` annotation. Here's a quick example of testing if our pose is changing as we drive. 
+
+The `fastForward` allows time to pass to give our drive time to move. 
+
+Then we call our `getPose()` method and compare it to a new Pose2d, which returns a value of 0. So this test ensures that whenever our drive method is called, the robot doesn't stay at the Pose2d but actually moves.
+
+```java
+  @Test
+  public void testPose() {
+    drive.drive(0.8, 0.8);
+    fastForward();
+
+    assertNotEquals(new Pose2d(), drive.getPose());
+  }
+```
+
+#### Running the tests
+
+Again, make sure to check out the WPILib [guide](https://docs.wpilib.org/en/stable/docs/software/wpilib-tools/robot-simulation/unit-testing.html) on this as it also talks about writing and running tests. To actually see if the tests are valid, open the WPILib command palette (top right corner logo) and search `Test Robot Code` and press enter. After it runs, it will show you the results of all your tests in the terminal.  
```

Comment:
ahhh did i mention i hate this guide. Also, if they have the vscode extension Test Runner for Java installed (which they should) they can just press the little green run button!